cmd/yggdrasil/main.go

```diff
@@ -361,6 +364,13 @@ func run(args yggArgs, ctx context.Context, done chan struct{}) {
 		logger.Errorln("An error occurred starting TUN/TAP:", err)
 	}
 	n.tuntap.SetupAdminHandlers(n.admin)
+	// Change user if requested
+	if args.chuser != "" {
```

Comment:
It may make sense to change to `nobody:nobody` by default, so things are a little less unsafe for people who just `sudo yggdrasil -autoconf` or similar. Thoughts @neilalexander?

Reply:
@Arceliar I think it should be noted that `nobody:nobody` is not standard across Linux distros. For example, Debian and friends use `nobody:nogroup` instead. We should check which group is available first or maybe use something like Dynamic Users in systemd (use a random UID/GID that's currently not in use by the system and reserve it).

Reply:
could just use `nobody` and maybe it would just resolve primary group on its own i think?
also would need to check UID being 0 before doing chuser by default (as this can be started as own user if right caps are given).
and opting out may be kinda uglier - manually specifying empty user string, though i can't really find many reasons of opting out, other than some possible exotic distributions/installations without proper user databases - which i guess i could imagine in some sterilized docker environments or routers but i don't really know how common, and probably checking existence of `nobody` user before deciding to default to chuser'ing to it would make sense - and we could just check database before setting default to `nobody`, likely writing function in chuser_unix.go for that (& stub returning empty string in chuser_other.go).
Requesting opinions for this mindfuzz of mine i guess

Reply:
I thought about this a bit more and probably programming in this kind of logic into daemon itself is too much of a magical behavior imo, and this should be provided in distributions' `.service` files instead.

Reply:
well there isn't any need for this to be in any distro, Debian has a pretty hardened default that runs as yggdrasil user from the start. I think all you need to get it to work is CAP_NET_ADMIN, I think this is how Debian configured their OpenVPN services too

Reply:
Non-systemd gentoo and alpine would benefit from this, as from my experience OpenRC scripts can't easily give out capabilities.

Reply:
Oh also *BSDs

Reply:
And [openwrt](https://github.com/openwrt/packages/blob/openwrt-21.02/net/yggdrasil/files/yggdrasil.init)

Reply:
It's worth mentioning that on Linux systems without systemd you could simply do: `sudo capsh --user='yggdrasil' --inh='cap_net_admin,cap_net_bind_service' --addamb='cap_net_admin,cap_net_bind_service' --shell=/bin/sh -- -c 'exec yggdrasil -autoconf'`

Reply:
I'm aware of capsh, and tried using it in past for exactly this purpose (which ended looking too much of a mess for my tastes), but evidently many people either don't know, or they don't like such an obscure & hacky looking line... https://git.alpinelinux.org/aports/tree/community/yggdrasil/yggdrasil.initd for example.
I also just tried it myself & it failed with `Admin socket failed to listen: listen unix /var/run/yggdrasil.sock: bind: permission denied` so it's not so simple i guess...

Reply:
Also as far as i can see in [yggdrasil.service](https://github.com/yggdrasil-network/yggdrasil-go/blob/develop/contrib/systemd/yggdrasil.service), it doesn't even specify `User`, so appears to run as root, at least in my archlinux system.

Reply:
You should be able to genconf and disable the unix socket/move it to a location yggdrasil user could write to

Reply:
Weird, I can't see creation of `/var/run/yggdrasil` or yggdrasil user (only group) in https://github.com/yggdrasil-network/yggdrasil-go/blob/develop/contrib/deb/generate.sh
Am I looking at wrong place?

Reply:
Debian being Debian maintains their own fork of yggdrasil with their own enhancements.

Reply:
I see. This packaging is indeed fairly good. This has gotten a bit offtopic, but it would be cool if one could specify default socket / config path at compile time (eg using `go build -ldflags="-X defaults.defaultAdminListen=unix:///var/run/yggdrasil/yggdrasil.sock"`) so that linux distros wouldn't need patches there to achieve saner packaging.
Either way this repo's systemd files suck at running non-root, and so consequentially do vendor (not official debian's) debs and archlinux package.

Reply:
On the whole, I'm very pro-having separate flags for UID and GID, because then it is clearer that we can do one without the other, and it is easier for packagers to do the right thing for their distribution/platform.

I'm not entirely sure whether we want to do that *by default* — we could certainly do some testing to see how it feels. The main thing I can think of that might break is setting up new multicast sockets when an interface disappears and reappears later but we've already dropped privileges, but maybe it'll be OK?

Reply:
Actually isn't that what #817 is doing?

Reply:
#817 can take only numeric UID/GIDs, which is unwieldy.
>because then it is clearer that we can do one without the other

why would you want do one but not another? everywhere i looked around, one either wants to set uid and also gid to primary group of uid, or uid/gid separate, which is also possible. with this implementation, one can also set gid without uid (eg `-user ":groupname"` or `-user ":gid"`), though it may be a little bit unobvious, but why would you want to not drop root?

if you think that instead of using `-user uid:gid` syntax we should do `-uid user -gid group`, i can change to that too, looks like a minimal change to me.

>The main thing I can think of that might break is setting up new multicast sockets when an interface disappears and reappears later but we've already dropped privileges, but maybe it'll be OK?

that's sorta fundamentally incompatible with privilege dropping tbh. on linux, we could probably do some magic with caps and reserve the right to deal with interfaces, but on other unixes we can't. restarting daemon when new interfaces to be managed pop up could probably make sense, but i don't quite feel that's what yggdrasil should do on its own either.

Reply:
unless multicast actually works without requiring root (i somehow confused it with managing interfaces, and using multicast aint that)